fedot/core/data/data_preprocessing.py

```diff
@@ -110,37 +108,45 @@ def force_categorical_determination(table):
     return categorical_ids, non_categorical_ids
 
 
-def data_has_missing_values(data: Union[InputData, 'MultiModalData']) -> bool:
+def data_has_missing_values(data: InputData) -> bool:
     """ Check data for missing values."""
-
-    if not isinstance(data, InputData):
-        for data_source_name, values in data.items():
-            if data_type_is_table(values):
-                return pd.DataFrame(values.features).isna().sum().sum() > 0
-    elif data_type_is_suitable_preprocessing(data):
+    # TODO: @andreygetmanov (add check for missing text data and write test)
+    if data_type_is_suitable_preprocessing(data):
         return pd.DataFrame(data.features).isna().sum().sum() > 0
     return False
 
 
-def data_has_categorical_features(data: Union[InputData, 'MultiModalData']) -> bool:
+def data_has_categorical_features(data: InputData) -> bool:
     """
     Check data for categorical columns.
     Return bool, whether data has categorical columns or not
     """
     if data.data_type is not DataTypesEnum.table:
         return False
-    data_has_categorical_columns = False
 
-    if not isinstance(data, InputData):
-        for data_source_name, values in data.items():
-            if data_source_name.startswith('data_source_table'):
-                features_types = values.supplementary_data.column_types.get('features')
-                cat_ids, non_cat_ids = find_categorical_columns(values.features, features_types)
-                data_has_categorical_columns = len(cat_ids) > 0
-
-    else:
-        features_types = data.supplementary_data.column_types.get('features')
-        cat_ids, non_cat_ids = find_categorical_columns(data.features, features_types)
-        data_has_categorical_columns = len(cat_ids) > 0
+    features_types = data.supplementary_data.column_types.get('features')
+    cat_ids, non_cat_ids = find_categorical_columns(data.features, features_types)
+    data_has_categorical_columns = len(cat_ids) > 0
 
     return data_has_categorical_columns
+
+
+def data_has_text_features(data: InputData) -> bool:
+    """
+    Check data for text fields.
+    Return bool, whether data has text fields or not
+    """
+
+    return isinstance(data.features[0], str) and len(data.features.shape) == 1
```

Comment:
А может тут просто обращаться к data_type? В InputData тут же есть такое поле, или есть причина, почему решил по другому определять текстовые / не текстовые данные

Reply:
> А может тут просто обращаться к data_type? В InputData тут же есть такое поле, или есть причина, почему решил по другому определять текстовые / не текстовые данные?

В текущем варианте (когда текст подаётся одним csv, табличные данные - другим) разницы особой действительно нет. Но одним из следующих PR планирую решить проблему загрузки "текст+таблица" одним csv (потому что текущий вариант страшно неудобный.  И вот тогда пригодится определение текстовых полей в таблице и связанные с этим методы

Reply:
Но ведь ячейки же могут содержать string и при этом не требовать обработки именно как текст. То есть я имею в виду, что в таблице вполне могут присутствовать категориальные признаки - и при этом это обычные табличные данные

Reply:
Ох, вот такой подход мне на самом деле не очень нравится - легко нарваться на данные, где предложения короткие, и это текст. Иди табличку, где названия категорий очень длинное. 

Ну как вариант можно смотреть сколько уникальных ячеек в таблице есть - если повторяющихся нет, то скорее всего имеем дело с текстом. Но, в целом, обсуждаемо: свистать всех сюда, Коля @nicl-nno, Гриша @gkirgizov, Валера @valer1435. Ваши идеи?

Reply:
Вообще, я сейчас мало представляю, как работает мультимодальность в федоте, особенно при работе с текстом. Вижу два возможных пути
1) Юзер сам задает - какие поля у него текст - юзеру больнее + нам надо сделать возможность прокидывать эти типы через input_data
2) Резко разделить сущности текста и таблицы - так как они обрабатываются совершенно по-разному и заставлять юзеру делать отдельные файлы с текстом и с таблицей - звучит разумно, так как это разные модальности - нечего им делать в одной  таблице (+ будет легче разделять логику работы с текстом, не смешивая ее в класс таблицы). Для аналогии можно привести такой пример - мы же не храним в одном csv файле картинку попиксельно и какую-то мета-информацию о ней.

Reply:
У меня есть идея сделать так:
Считать текстовыми колонками только те, где есть префикс text_ (всё-таки такие кейсы не слишком часты). 
Далее преобразовывать их в отдельный текстовый datasource, удаляя этот признак из табличного data source.

Reply:
> Вообще, я сейчас мало представляю, как работает мультимодальность в федоте, особенно при работе с текстом. Вижу два возможных пути

Если говорить про первый путь, то сам о чём-то подобном думал. Вообще в целом чувствую, что есть необходимость дополнить класс MultiModalData методами импорта из разных файлов (сейчас это всё делается при помощи InputData и потом собирается вместе. В целом реализация кажется не такой сложной, а задавать текстовые колонки при импорте юзер тоже сможет.

Если говорить про второй, он сейчас используется. И я понял, насколько это неудобно, когда попробовал запустить сsv с текстовыми полями. Это надо извлечь из csv текстовые поля в отдельный объект, потом извлечь табличные в другой, потом их объединить в MultiModalData. Это правда страшно неудобно. Многие мультимодальные датасеты совмещают текст и таблицу в одном файле, потому что текст это не картинка - текст можно засунуть в csv, и это часто делают.

Reply:
> У меня есть идея сделать так: Считать текстовыми колонками только те, где есть префикс text_ (всё-таки такие кейсы не слишком часты). Далее преобразовывать их в отдельный текстовый datasource, удаляя этот признак из табличного data source.

Получается, если префикса не будет, у юзера не будет возможности задать определённую колонку как текст? Просто тогда ему придётся переименовывать колонку, что тоже звучит неудобно

Reply:
> Ох, вот такой подход мне на самом деле не очень нравится - легко нарваться на данные, где предложения короткие, и это текст. Иди табличку, где названия категорий очень длинное.


Согласен


> Ну как вариант можно смотреть сколько уникальных ячеек в таблице есть - если повторяющихся нет, то скорее всего имеем дело с текстом

Вот это, как мне кажется, достаточно хороший способ отсечь текст от категориальных признаков

Reply:
У меня такие мысли в ответ на предложения @valer1435 и @nicl-nno

> заставлять юзеру делать отдельные файлы с текстом и с таблицей
Это может быть не очень удобно. Аналогия с картинками не очень рабочая, так как в ходу, насколько я знаю, нет таких форматов данных, где описание+картинка в чем-то вроде одного csv. С другой стороны, легко можно представить табличные данные + текст: любая выгрузка какой-нибудь базы данных (например, текстовые запросы к поисковой система с метаданным) даст нам такую мультимодальную табличку. 

> Юзер сам задает - какие поля у него текст - юзеру больнее + нам надо сделать возможность прокидывать эти типы через input_data
Такой способ явного указания точно стоит предусмотреть. Но и автоматическое определение в случае неуказания может быть удобным -- см. пример из предыдущего поинта. Если есть возможность более менее разумно автоматичеки определить данные и это упростить предобработку данных пользователю -- почему бы и нет? Таки ж мы неспроста *АвтоМэЛэ*.

> Считать текстовыми колонками только те, где есть префикс text_ (всё-таки такие кейсы не слишком часты).
В связи с пред. пунктом -- лишние действия для пользователя, если у него уже готовая табличка с данными. Да и дизайн с конвенциями именования -- в целом не лучшая идея на долгий срок (because naming conventions are difficult to enforce & check).

Мои мысли насчет автоопределения того, текстовые ли данные у нас в столбце -- это развитие идеи @Dreamlone:
> Ну как вариант можно смотреть сколько уникальных ячеек в таблице есть - если повторяющихся нет, то скорее всего имеем дело с текстом.
С учетом того, что и в текстовых данных могут быть дубликаты, я бы предложил считать так: если *доля уникальных значений меньше какого-то threshold (0.1? 0.05?)* **ИЛИ** *количество уникальных меньше какого-то N (10?)* то это категориальный признак. В противном случае строки -- это

Reply:
> С учетом того, что и в текстовых данных могут быть дубликаты, я бы предложил считать так: если _доля уникальных значений меньше какого-то threshold (0.1? 0.05?)_ **ИЛИ** _количество уникальных меньше какого-то N (10?)_ то это категориальный признак. В противном случае строки -- это текст.

Т.к. эта задача на следующий PR, могу попутно провести мини-ресёрч и позапускать мультимодальные бенчмарки, чтобы определить этот threshold более осознанно

Со всем остальным согласен!

Reply:
> > _количество уникальных меньше какого-то N (10?)_ то это категориальный признак. В противном случае строки -- это текст.


Хотя здесь не факт, что в большом датасете не окажется условно 50 уникальных признаков

Reply:
Я за идею Гриши с предложением считать относительный порог уникальности. Ресерча тут наверно даже не требуется - просто нужно волевым решением выбрать логичное значение - в любом случае это будет компромисс. Пусть, например, если доля уникальных значений больше 95% процентов, то это текст.

Однако **важно** определить место, в котором мы это дело будем определять: в препроцессоре примерно по такому же принципу обрабатываются `string` в таблицах (вспоминаем наш недавний [пост](https://habr.com/ru/company/ods/blog/657525/)) - если количество уникальных категорий велико, то препроцессор попытается перевести столбец в numeric тип, и, если не выйдет, то он просто выкинет столбец из таблицы. Значит надо подготовить абстракцию, которая будет заранее парсить табличку и определять а к какому DataType относится каждая её часть. Но как и писал выше, мы вполне спокойно можем нарваться на случай `['1.0', '2.5', '3.1', ..., '4.1']`, когда вроде и строки и вроде и количество уникальных значений велико, а вот на деле это просто float столбец (и хорошо его расколбасит, когда мы попытаемся обработать его как text).

В любом случае в рамках данного PR предлагаю эту проблему не решать, вполне можно оформить issue и сослаться в нем на наше обсуждение здесь. @andreygetmanov, сделай плиз, а пока постарайся пожалуйста как-нибудь обойти проблему с автоматическим определением типов данных в текущем PR

Reply:
> В любом случае в рамках данного PR предлагаю эту проблему не решать, вполне можно оформить issue и сослаться в нем на наше обсуждение здесь. @andreygetmanov, сделай плиз, а пока постарайся пожалуйста как-нибудь обойти проблему с автоматическим определением типов данных в текущем PR

Issue [оформил](https://github.com/nccr-itmo/FEDOT/issues/754). Не совсем понимаю, что ты имеешь в виду под проблемой, потому что в текущем варианте данные считаются текстовыми, если в датасете всего 1 столбец и он имеет тип str (т.к. мы текстовый data source подаём отдельным файлом). Текущая проверка, кажется, должна работать достаточно хорошо

Reply:
Давай разбираться. Дано: 
Столбец, в котором каждая ячейка содержит текстовый тип данных (`string`), но при этом выглядит вот так
`['1.0', '2.5', '3.1', ..., '4.1']`. Тот алгоритм, который ты описал, вполне спокойно её пометит как текст, так как все ячейки строки и при этом количество уникальных значений велико. 

Еще мы не затрагиваем тут момент у предложенной проверки, что будет, если столбец с текстовыми данными, а некоторые ячейки содержат nan'ы, которые воспринимаются как float - это сразу будет не текстовый столбец?

Reply:
> Давай разбираться. Дано: Столбец, в котором каждая ячейка содержит текстовый тип данных (`string`), но при этом выглядит вот так `['1.0', '2.5', '3.1', ..., '4.1']`. Тот алгоритм, который ты описал, вполне спокойно её пометит как текст, так как все ячейки строки и при этом количество уникальных значений велико.
> 
> Еще мы не затрагиваем тут момент у предложенной проверки, что будет, если столбец с текстовыми данными, а некоторые ячейки содержат nan'ы, которые воспринимаются как float - это сразу будет не текстовый столбец?


Согласен со всем, что после "дано", но я как раз о том, что дано. Сейчас чтобы автоматически установить тип данных как text, надо подать на вход не больше одного столбца признаков. Я исходил из логики, что при попытке что-то предсказать на одном столбце табличных данных возможное ложное определение его типа как text - это далеко не самая большая проблема
Могу отключить пока проверку содержимого данных и просто чекать data_type == text. Или можно как-то ещё поправить?

Reply:
А, если столбец один, то вполне себе рабочий вариант. Я то просто думал, что ты парсишь таблицу на текстовые и не текстовые столбцы. Оставь как есть, только напиши плиз тест на такой случай, чтобы проверить, что всё ок отрабатывает как ты ожидал

Reply:
> напиши плиз тест на такой случай, чтобы проверить, что всё ок отрабатывает как ты ожидал

А [этот](https://github.com/nccr-itmo/FEDOT/blob/865b3db19d87f5d657bfbf4b7b95a3d40b4b4797/test/unit/preprocessing/test_preprocessing_though_api.py#L199) разве не покрывает?

Reply:
Отчасти, но кмк стоит побольше потестировать на разных примерах работу препроцессора с автоматическим определением типа данных. Ну а по поводу граничных случаев, можно взять например: 
1. `['adsdfksf;ksd;fksf', 'dsfsd;skf;sk', 'fdsfsdafs', ..., 'fdfgadssd']` - только текст (уже реализовано)
2. `[np.nan, 'adsdfksf;ksd;fksf', 'fdsfsdafs', ..., 'fdfgadssd']` - есть и текст, и nan на первом месте в массиве
3. `['1.5', np.nan, '2', ..., '4.2']` - есть и текст, и nan, но на самом деле данные float (или что-то типо такого)

Вполне возможно, что один из тестируемых сценариев подскажет, что определять тип данных по первому элементу в массиву неоптимально.